services/QuillLMS/spec/models/unit_template_spec.rb

```diff
@@ -204,6 +204,91 @@
+    let!(:student_three) { create(:student) }
+    let!(:student_four) { create(:student) }
+    let!(:classroom_one) { create(:classroom, students: [student_one, student_two, student_three]) }
+    let!(:classroom_two) { create(:classroom, students: [student_four]) }
+    let!(:unit_one) { create(:unit) }
+    let!(:unit_two) { create(:unit) }
+    let!(:classroom_unit_one) { create(:classroom_unit, unit: unit_one, classroom: classroom_one, assigned_student_ids: classroom_one.students.ids) }
+    let!(:classroom_unit_two) { create(:classroom_unit, unit: unit_two, classroom: classroom_two, assigned_student_ids: classroom_two.students.ids) }
+
+    it 'returns the expected total and assigned student counts' do
+      classrooms = [classroom_one, classroom_two]
+      unit_one_counts = UnitTemplate.student_counts_for_previously_assigned_activity(unit_one, unit_one.classrooms)
+      unit_two_counts = UnitTemplate.student_counts_for_previously_assigned_activity(unit_two, unit_two.classrooms)
+      expect(unit_one_counts).to eq([
+        {
+          total_student_count: classroom_one.students.length,
+          assigned_student_count: classroom_unit_one.assigned_student_ids.length
+        }
+      ])
+      expect(unit_two_counts).to eq([
+        {
+          total_student_count: classroom_two.students.length,
+          assigned_student_count: classroom_unit_two.assigned_student_ids.length
+        }
+      ])
+    end
+  end
+
+  describe '#previously_assigned_activity_data' do
+    let!(:current_user) { create(:teacher_with_a_couple_classrooms_with_one_student_each) }
+    let!(:classroom) { current_user.classrooms_i_teach.first }
+    let!(:unit_one) {create(:unit, user_id: current_user.id)}
+    let!(:unit_two) {create(:unit, user_id: current_user.id)}
+    let!(:activity_one) { create(:activity) }
+    let!(:activity_two) { create(:activity) }
+    let!(:activity_three) { create(:activity) }
+    let!(:activity_four) { create(:activity) }
+    let!(:activity_five) { create(:activity) }
+    let!(:classroom_unit_one) { create(:classroom_unit, unit: unit_one, classroom: classroom, assigned_student_ids: classroom.student_ids) }
+    let!(:classroom_unit_two) { create(:classroom_unit, unit: unit_two, classroom: classroom) }
+    let!(:unit_activity_one) { create(:unit_activity, unit: classroom_unit_one.unit, activity: activity_one) }
+    let!(:unit_activity_two) { create(:unit_activity, unit: classroom_unit_one.unit, activity: activity_two) }
+    let!(:unit_activity_three) { create(:unit_activity, unit: classroom_unit_two.unit, activity: activity_two) }
+    let!(:unit_activity_four) { create(:unit_activity, unit: classroom_unit_two.unit, activity: activity_three) }
+    let!(:activity_ids) { [activity_one.id, activity_two.id, activity_three.id, activity_four.id, activity_five.id] }
+    let!(:count) { classroom.student_ids.length }
+    let!(:results) { UnitTemplate.previously_assigned_activity_data(activity_ids, current_user) }
```

Comment:
Then you could rewrite results in a way that gets rid of some of the duplication in the specs below:

`let!(:results) { subject[:previously_assigned_activity_data] }`